File: 1-two-sum/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // Create an array of custom data class to store values and their original indices
        Data[] array = new Data[nums.length];
        
        // Fill the array with values and their corresponding indices
        for (int i = 0; i < nums.length; i++) {
            array[i] = new Data(nums[i], i);
        }
        
        // Sort the array based on the value
        Arrays.sort(array, (a, b) -> Integer.compare(a.val, b.val));
        
        int i = 0, j = array.length - 1;
        while (i < j) {
            int sum = array[i].val + array[j].val;
            if (sum == target) {
                // Return the indices of the original positions
                return new int[] { array[i].idx, array[j].idx };
            } else if (sum < target) {
                i++;
            } else {
                j--;
            }
        }
        
        return null;
    }
    
    // Data class to store value and its index
    static class Data {
        int val;
        int idx;
        
        Data(int val, int idx) {
            this.val = val;
            this.idx = idx;
        }
    }
}

File: 1-two-sum/two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let hashTable = {};  // Use an object to mimic a hash table
    
    for (let i = 0; i < nums.length; i++) {
        let complement = target - nums[i];
        
        // Check if the complement exists in the hash table
        if (hashTable[complement] !== undefined) {
            return [hashTable[complement], i];  // Return the indices of the two numbers
        }
        
        // Store the current number and its index in the hash table
        hashTable[nums[i]] = i;
    }
    
    // Return an empty array if no solution is found (although per the problem statement, one solution always exists)
    return [];
};
